### PR TITLE
Use Optimisers.jl

### LINKED ISSUE #637: New New Optimisers
If we go full steam ahead with more functional-style AD (#628), then we'll need to rework optimisers a little. I think at the core the update step will look very similar, if a bit more functional, something like:

```julia
apply(::Optimiser, x, dx, state = ...) -> (dx', state')
```

Then we have an `update` function which actually applies the gradient. In general, the `x`, `dx` and `state` can be structs and we'll rebuild `x` by recursing over it (somewhat like mapleaves, though we won't need to depend on `treelike` since we can just use reflection here).

I think this is all reasonably straightforward; the main wrinkle in my mind is how we enable in-place updates as an optimisation (since it's not ideal to have two copies of ResNet at once). I'm not aware of any great solution to this right now, so we might need to define an `ismutable` trait and declare it for types we care about.

### LINKED ISSUE #823: Flux Optimizers should define equality
I was writing some tests, and I wanted to be sure that the hyper-parameters of two models were the same.
So I was checking equality on an object that contained all my hyperparameters,
including the optimizer (which has its own hyper-parameters).

Turns out that fails as
```
julia> Flux.Optimise.NADAM(0.003, (0.9, 0.9), IdDict{Any,Any}()) == Flux.Optimise.NADAM(0.003, (0.9, 0.9), IdDict{Any,Any}())
false
```

### PR DESCRIPTION
**Pull request by DhairyaLGandhi**
Fixes #637, fixes #823.

### CODE REVIEW

**Comment by DhairyaLGandhi** [comment]
@ModelZookeeper commands

File: src/optimise/train.jl [hunk 1]
```diff
@@ -116,6 +117,43 @@ function train!(loss, ps, data, opt; cb = () -> ())
   end
 end
 
+function train!(m, loss, data, opt; cb = (x...) -> ()
+                                    prehook = (x...) -> (),
+                                    posthook = (x...) -> ())
+  st = [Optimisers.init(opt, p) for p in Flux.params(m)]
```
Comment (ToucheSir) [inline src/optimise/train.jl:123]: I think this can be reduced to `st = Optimisers.state(opt, m)`.

Comment (darsnack) [reply src/optimise/train.jl:123]: I think this should be defined as a function in the `optimiser.jl` file. One for `state(opt, p::Params)` and another for `update(opt, x::Params, dx, state)`.

**Comment by darsnack** [comment]
I would prefer we remove the hooks from `Flux.train!` in this PR and save it for something else. This PR can simply remove the optimizers in favor of Optimisers.jl.

**Comment by DhairyaLGandhi** [comment]
Done a bunch of cleanup, including removing the train loop changes, but I feel like we would circle back to it soon

File: src/optimise/train.jl [hunk 2]
```diff
@@ -97,12 +102,13 @@ Multiple optimisers and callbacks can be passed to `opt` and `cb` as arrays.
 function train!(loss, ps, data, opt; cb = () -> ())
   ps = Params(ps)
   cb = runall(cb)
+  st = [Optimisers.init(opt, p) for p in ps]
```
Comment (darsnack) [inline src/optimise/train.jl:105]: Could we define `init` for `Params` like we do `update`?

Comment (DhairyaLGandhi) [reply src/optimise/train.jl:105]: I'm thinking if this should be wholesale replaced with `Optimisers.state` instead

Comment (DhairyaLGandhi) [reply src/optimise/train.jl:105]: I'm only worried about corner cases that this might hit

Comment (darsnack) [reply src/optimise/train.jl:105]: I think `state` makes sense. For every `p` in `ps`, we should be able to independently optimize each one by calling the user facing `state` and `update`. I think its safe to do that here.

Comment (DhairyaLGandhi) [reply src/optimise/train.jl:105]: Right, so that works in cases where we have arrays as params, but not arbitrary structs, which is what we want.

The user facing API is something like (pseudocode; although it might work in this MWE)
```julia
m = Dense(3,3)
opt = ADAM()
st = Optimisers.state(opt, m) # `m` could contain arbitrary structs which shouldn't be functor'd
loss(m, x, y) = Flux.mse(m(x), y)
for i = 1:1000
  gs, = gradient(m) do m
    @show loss(m, w, w′)
  end
  m, st = opt(m, gs, st)
end
```

Comment (darsnack) [reply src/optimise/train.jl]: That's exactly what I'm suggesting. To be more specific, change the code to:
```suggestion
  st = state(opt, ps)
```

And elsewhere, define:
```julia
init(o, ps::Params) = [init(o, p) for p in ps]
```

Should allow for the future case where `m` is not a `Params` but also the current case where we need to support `Params`.

File: docs/src/training/optimisers.md [hunk 3]
```diff
@@ -97,9 +98,9 @@ v = ρ * v - η * Δ
 w = w - v
 ```
 
-The `apply!` defines the update rules for an optimiser `opt`, given the parameters and gradients. It returns the updated gradients. Here, every parameter `x` is retrieved from the running state `v` and subsequently updates the state of the optimiser.
+The `apply` defines the update rules for an optimiser `opt`, given the parameters and gradients. It returns the updated gradients. Here, every parameter `x` is retrieved from the running state `v` and returns the new state of the optimizer.
```
Comment (darsnack) [inline docs/src/training/optimisers.md:101]:
```suggestion
The `apply` defines the update rules for an optimiser `opt`, given the parameters and gradients. It returns the updated gradients and optimizer state.
```

File: src/optimise/train.jl [hunk 4]
```diff
@@ -19,15 +19,20 @@ according to optimizer `opt`  and the gradients `gs` (the gradient `g`).
 
 As a result, the parameters are mutated and the optimizer's internal state may change.
 """
-function update!(opt, x, x̄)
-  x .-= apply!(opt, x, x̄)
+function update!(opt, x, x̄, st)
+  x̄, st = apply(opt, x, x̄, st)
+  x .-= x̄
```
Comment (darsnack) [inline src/optimise/train.jl:24]: Should we use `Optimisers.patch` here for consistency?

File: src/optimise/train.jl [hunk 5]
```diff
@@ -19,15 +19,20 @@ according to optimizer `opt`  and the gradients `gs` (the gradient `g`).
...
-  for x in xs
+function update!(opt, xs::Params, gs, st)
+  st_ = []
+  for (x,s) in zip(xs,st)
     gs[x] == nothing && continue
-    update!(opt, x, gs[x])
+    _, s = update!(opt, x, gs[x], s)
+    append!(st_, s)
   end
+  nothing, st_
```
Comment (darsnack) [inline src/optimise/train.jl:35]:
```suggestion
  xs, st_
```

File: src/optimise/train.jl [hunk 6]
```diff
@@ -97,12 +104,13 @@ Multiple optimisers and callbacks can be passed to `opt` and `cb` as arrays.
 function train!(loss, ps, data, opt; cb = () -> ())
   ps = Params(ps)
   cb = runall(cb)
+  st = Optimisers.init(opt, ps)
```
Comment (darsnack) [inline src/optimise/train.jl:107]:
```suggestion
  st = Optimisers.state(opt, ps)
```

**Review by darsnack — COMMENTED** [review]
Looking pretty good. Two ML calls ago, we talked about having `apply!` also defined in Optimisers.jl as a temporary measure to avoid potential performance regressions. Is that still the plan?

**Comment by DhairyaLGandhi** [comment]
See https://github.com/FluxML/Optimisers.jl/pull/13

We would also want #1402 i think when walking through the tests

**Comment by darsnack** [comment]
Cool, looks good. Should we be calling `Optimisers.update` here then? Calling `apply` directly bypasses the mutability check.

**Comment by darsnack** [comment]
Ref https://github.com/FluxML/Flux.jl/pull/1613#issuecomment-863274729

In my mind, this PR is basically good to go. What I would suggest is establishing a `optimizers.jl` branch off `master` (similar to how we had a `zygote` branch for that transition). Then we can safely merge this into the `optimisers.jl` branch without much concern, and everyone can start the process of benchmarking/validating the transition.

**Comment by DhairyaLGandhi** [comment]
That branch exists already - it's this pr! I think we're kind of tied to merge the zeros first, then this and boom.

**Comment by darsnack** [comment]
Why does this depend on `Zeros`?

If this is _the_ branch, then I guess let's rebase it?

**Comment by DhairyaLGandhi** [comment]
We can rebase. Without the zeros, state initialisation and optimisation doesn't work. In order to define methods in init, one would have to do it manually.

**Comment by darsnack** [comment]
I'm not sure what you mean. Why is this not enough
```julia
Optimisers.init(o, ps::Params) = [Optimisers.state(o, p) for p in ps]
```

**Comment by DhairyaLGandhi** [comment]
We also have a route that does not require the use of `Params`.

**Comment by darsnack** [comment]
You mean where we allow passing in `[W, b]`? Wouldn't that route be handled by Optimisers.jl already?

**Comment by CarloLucibello** [comment]
Closing as stale: this PR has not seen any new commits in over two years, and Flux has changed substantially since then (Optimisers.jl-based training, the new recurrent layers, Reactant/Enzyme support, etc.), so it can no longer be merged as-is.

This is housekeeping, not a rejection of the idea. If you (or anyone else) would like to pick it up, please feel free to reopen this PR or open a fresh one against the current `master` — the contribution is still appreciated.
